### PR TITLE
add child workflow support to test suite

### PR DESCRIPTION
**Pull request by yiminc**
Adding child workflow to test suite.

Known issue: need to rethink about how we auto forward the mock clock. The clock probably does not work in some cases with child workflow. I plan to address that in a separate diff.

### CODE REVIEW

File: internal_workflow_testsuite.go [hunk 1]
```diff
@@ -296,6 +364,26 @@ func (env *testWorkflowEnvironmentImpl) overrideActivity(activityFn, fakeActivit
 	env.overrodeActivities[fnName] = fakeActivityFn
 }
 
+func (env *testWorkflowEnvironmentImpl) overrideWorkflow(workflowFn, fakeWorkflowFn interface{}) {
+	// verify both functions are valid workflow func
```
Comment (venkat1109) [inline internal_workflow_testsuite.go]: this is the same code for overrideActivity() - consider refactoring the validation into its own helper - something like validateFunctionOverride(actual, fake) string

File: internal_workflow_testsuite.go [hunk 2]
```diff
@@ -827,13 +936,56 @@ func (env *testWorkflowEnvironmentImpl) RegisterSignalHandler(handler func(name
...
+		// check if current workflow is a child workflow
+		parentEnv := env.parentEnv
+		if parentEnv != nil && parentEnv.onChildWorkflowCanceledListener != nil {
+			parentEnv.postCallback(func() {
+				parentEnv.onChildWorkflowCanceledListener(env.workflowInfo)
+			}, false)
+		}
+	} else if childHandle, ok := env.childWorkflows[workflowID]; ok {
+		// current workflow is a parent workflow, and we are canceling a child workflow
+		delete(env.childWorkflows, workflowID)
```
Comment (venkat1109) [inline internal_workflow_testsuite.go]: shouldn't this be deleted inside the if{} instead of the else{} ?

File: internal_workflow_testsuite.go [hunk 3]
```diff
@@ -186,6 +211,38 @@ func newTestWorkflowEnvironmentImpl(s *WorkflowTestSuite) *testWorkflowEnvironme
 	return env
 }
 
+func (env *testWorkflowEnvironmentImpl) newTestWorkflowEnvironmentForChild(options *workflowOptions) *testWorkflowEnvironmentImpl {
+	// create a new test env
+	childEnv := newTestWorkflowEnvironmentImpl(env.testSuite)
+	childEnv.parentEnv = env
```
Comment (venkat1109) [inline internal_workflow_testsuite.go]: add this workflow to the parent's map[childWorkflows]

File: internal_workflow_testsuite.go (hunk 2 above)
Comment (yiminc) [inline internal_workflow_testsuite.go]: no, it should be in the else case, as we are deleting child from parent, so we do it when current workflow is parent workflow.

File: internal_workflow_testsuite.go (hunk 1 above)
Comment (yiminc) [inline internal_workflow_testsuite.go]: will update

File: internal_workflow_testsuite.go (hunk 3 above)
Comment (yiminc) [inline internal_workflow_testsuite.go]: child was added in ExecuteChildWorkflow() method. Put it here also works, will update.

**Pull request merged**